### PR TITLE
Upcoming Event Search

### PR DESCRIPTION
**Pull request by hilal-demir**

### CODE REVIEW

File: Platon/backend/tests/search_engine_test.py [hunk 1]
```diff
@@ -264,3 +265,78 @@ def test_institution_search(self):
         self.assertEqual(200,actual_response.status_code)
         self.assertEqual(expected_result,json.loads(actual_response.data))
 
+
+class UpcomingEventsSearchTest(BaseTest):
```
Comment (cakmadam98) [inline Platon/backend/tests/search_engine_test.py:269]: I think you forgot to include UpcomingEventsSearchTest into test.py folder.

Comment (cakmadam98) [reply Platon/backend/tests/search_engine_test.py:269]: When I tried to run the your test, I got the following error:

FAIL: test_deadline_search (tests.search_engine_test.UpcomingEventsSearchTest)
Traceback (most recent call last):
  File "/Users/cakmadam98/Desktop/4.1/CmpE451/bounswe2020group7/Platon/backend/tests/search_engine_test.py", line 342, in test_deadline_search
    self.assertEqual(expected_result, json.loads(actual_response.data))
AssertionError: {'num[42 chars]m': 'ANIVAE 2021', 'date': 'Mar 28, 2021 - Mar[260 chars])'}]} != {'num[42 chars]m': 'ISC 2021', 'date': 'Jun 2, 2021 - Jun 4, [606 chars])'}]}
Diff is 1188 characters long. Set self.maxDiff to None to see it.

**Review by cakmadam98 — CHANGES_REQUESTED** [review]

File: Platon/backend/app/search_engine/forms.py [hunk 2]
```diff
@@ -34,4 +34,15 @@ class WorkspaceSearchForm(Form):
...
 ws_search_parser.add_argument('per_page',type=int,help="Number of Records in a Page",location='args')
-ws_search_parser.add_argument('auth_token', type=str,help="Authantication Token(If registered)",location='headers')
\ No newline at end of file
+ws_search_parser.add_argument('auth_token', type=str,help="Authantication Token(If registered)",location='headers')
+
+class UpcomingEventsSearchForm(Form):
+    search_query = StringField("search_query",validators=[validators.DataRequired()])
+    page = IntegerField("page")
+    per_page = IntegerField("per_page")
+
```
Comment (HUmutOzdemir) [inline Platon/backend/app/search_engine/forms.py:43]: 1.1.6.2 Users and guests shall be able to do an advanced search on only public workspaces, upcoming events, and any profile by specifying filters such as founder of the project, starting date, submission deadline, research area/topic and labels.

According to this requirement, there must be a filtering mechanism on Upcoming Event Search which filters the events using their deadlines and starting dates. Can you add such a feature?

File: Platon/backend/app/search_engine/views.py [hunk 3]
```diff
@@ -308,7 +327,77 @@ def get(self):
...
+            tokens = SearchEngine.semantic_related_list(tokens)
+            # List of user dictionaries that will be send as a search output
+            result_list = []
+            # Score of each user record
+            result_id_score = []
+            for token, score in tokens:
+                try:
+                    query = '(LOWER(id) REGEXP ".*{0}.*" OR LOWER(title) REGEXP ".*{0}.*" \
+                    OR LOWER(acronym) REGEXP ".*{0}.*" OR LOWER(location) REGEXP ".*{0}.*") \
+                    OR LOWER(date) REGEXP ".*{0}.*" OR LOWER(deadline) REGEXP ".*{0}.*" OR LOWER(link) REGEXP ".*{0}.*"'  \
```
Comment (HUmutOzdemir) [inline Platon/backend/app/search_engine/views.py:358]: Why do you search for the ID of the event? Actually, I think ID has no meaning for the users that will use this endpoint. If you have a valid reason please state your reason.

File: Platon/backend/app/search_engine/views.py [hunk 4]
```diff
@@ -308,7 +327,77 @@ def get(self):
...
+                            result_id_score[id_index] = (upcoming_event[0], result_id_score[id_index][1] + score)
+                except:
+                    return make_response(jsonify({"error": "Database Connection Problem."}), 500)
+
+                # Sort result ids according to their scores
+                sorted_id_list = SearchEngine.sort_ids(result_id_score)
+                sorted_result_list = []
+                for upcoming_event_id, score in sorted_id_list:
+                    index = [upcoming_event["id"] for upcoming_event in result_list].index(upcoming_event_id)
+                    sorted_result_list.append(result_list[index])
```
Comment (HUmutOzdemir) [inline Platon/backend/app/search_engine/views.py:382]: 1.1.6.4 Users and guests should be able to sort search results of upcoming events according to the sorting criteria as date, alphabetical order.

According to this requirement, results shall be able to sort according to their dates and alphabetical order. Can you add such a feature?

File: Platon/backend/app/search_engine/views.py (hunk 3 above)
Comment (HUmutOzdemir) [reply Platon/backend/app/search_engine/views.py:358]: I think the link also has no meaning to the user.

**Review by HUmutOzdemir — CHANGES_REQUESTED** [review]

**Comment by HUmutOzdemir** [comment]
> _No description provided._

Please also merge your code firstly to the backend-dev branch. So we can try the code to see whether it creates a problem in other parts of the system or not. Then after trying the code on backend-dev we can merge it to the release branch.

File: Platon/backend/app/search_engine/forms.py (hunk 2 above)
Comment (hilal-demir) [reply Platon/backend/app/search_engine/forms.py]: Added them on my last commit.

File: Platon/backend/app/search_engine/views.py (hunk 4 above)
Comment (hilal-demir) [reply Platon/backend/app/search_engine/views.py]: Added them too.

File: Platon/backend/app/search_engine/views.py (hunk 3 above)
Comment (hilal-demir) [reply Platon/backend/app/search_engine/views.py]: I removed them.

**Review by cakmadam98 — CHANGES_REQUESTED** [review]

File: Platon/backend/app/search_engine/forms.py [hunk 5]
```diff
@@ -34,4 +34,21 @@ class WorkspaceSearchForm(Form):
...
 ws_search_parser.add_argument('page',type=int,help="Page ID(0-Indexed)",location='args')
 ws_search_parser.add_argument('per_page',type=int,help="Number of Records in a Page",location='args')
-ws_search_parser.add_argument('auth_token', type=str,help="Authantication Token(If registered)",location='headers')
\ No newline at end of file
+ws_search_parser.add_argument('auth_token', type=str,help="Authantication Token(If registered)",location='headers')
+
+class UpcomingEventsSearchForm(Form):
+    search_query = StringField("search_query",validators=[validators.DataRequired()])
+    date_filter = StringField("date_filter")
+    deadline_filter = StringField("deadline_filter")
```
Comment (HUmutOzdemir) [inline Platon/backend/app/search_engine/forms.py:42]: Why these 2 are strings and how do you filter these fields by only using 1 value? It is not user friendly to filter the dates by only using 1 value. I think it will be better to use 2 date values to represent an interval for date filtering.
Also, I can not understand why these 2 are StringFileds. I think you should take input by using DateTimeFields to make the format of dates uniform.

File: Platon/backend/app/search_engine/helpers.py [hunk 6]
```diff
@@ -99,7 +99,11 @@ def remove_punctuation(search_query):
 
     @staticmethod
     def sort_ids(related_id_score_list):
-        return sorted(related_id_score_list, key=lambda tup: tup[1])
+        return sorted(related_id_score_list, key=lambda tup: tup[1],reverse=True)
+
+    @staticmethod
+    def sort_results(results_list, key_list, reverse):
+        return sorted(results_list, key=lambda x: " ".join([x[key].lower() for key in key_list]), reverse=reverse)
```
Comment (HUmutOzdemir) [inline Platon/backend/app/search_engine/helpers.py]: These 2 are the functions that I implemented. Why did you change them? If you changed them user search will not be able to sort the results according to their relevance or alphabetical order.

File: Platon/backend/app/search_engine/views.py [hunk 7]
```diff
@@ -308,7 +329,114 @@ def get(self):
...
+                for result in result_list:
+                    event_date = result.get("date")
+                    event_date = SearchEngine.remove_punctuation(event_date)
+                    remove_digits = str.maketrans('', '', digits)
+                    filter_data = form.date_filter.data.translate(remove_digits)
+                    filter_data = SearchEngine.remove_punctuation(filter_data)
+                    if filter_data in event_date:
+                        date_filter_list.append(result)
+                    result_list = date_filter_list
+            sorted_result_list = []
```
Comment (HUmutOzdemir) [inline Platon/backend/app/search_engine/views.py:404]: I really can not understand your filtering algorithm. And I the,nk this part does not apply meaningful filtering to dates. Please update the date filtering part by converting strings to datetime objects by using the approach on the following link:
https://www.educative.io/edpresso/how-to-convert-a-string-to-a-date-in-python

File: Platon/backend/app/search_engine/views.py [hunk 8]
```diff
@@ -308,7 +329,114 @@ def get(self):
...
+                    sorted_result_list = SearchEngine.sort_results(result_list, ["title"], False)
+                    result_list = sorted_result_list
+                else:
+                    dates = []
+                    for result in result_list:
+                        date = result.get("date")
+                        if date != "N/A":
+                            dates.append(date)
+
+                    dates.sort(key = lambda date: datetime.strptime(date[0:12].strip(), '%b %d, %Y'))
```
Comment (HUmutOzdemir) [inline Platon/backend/app/search_engine/views.py:418]: Why didn't you use the part that I implemented on user search which sorts the results according to their relevance to the search query?

File: Platon/backend/app/search_engine/helpers.py [hunk 9]
```diff
@@ -101,6 +101,10 @@ def remove_punctuation(search_query):
     def sort_ids(related_id_score_list):
         return sorted(related_id_score_list, key=lambda tup: tup[1])
 
+    @staticmethod
+    def sort_results(results_list, key_list, reverse):
+        return sorted(results_list, key=lambda x: " ".join([x[key].lower() for key in key_list]), reverse=reverse)
+
```
Comment (HUmutOzdemir) [inline Platon/backend/app/search_engine/helpers.py:107]: Why did you change the function that I implemented? I used this function on user search to sort the results by using their alphabetical order. If you changed this function, my code will not be able to sort the users according to alphabetical order.

**Review by HUmutOzdemir — CHANGES_REQUESTED** [review]
I see some problems with your code. If you want to merge it otherwise, please delete the filtering part. Because it does not have the right functionality. Also please add the sorting part for relevance that I used in the user search.

File: Platon/backend/app/search_engine/views.py [hunk 10]
```diff
@@ -433,7 +452,124 @@ def get(self):
...
+                    return make_response(jsonify({"error": "Database Connection Problem."}), 500)
 
+            # Search tokens for deadline match
+            deadline_filter_start_list = []
+            if form.deadline_filter_start.data is not None:
+                for result in result_list:
+                    event_deadline = result.get("deadline")
+                    filter_data = form.deadline_filter_start.data
+                    if datetime.strptime(event_deadline, '%b %d, %Y') >= filter_data:
+                        deadline_filter_start_list.append(result)
```
Comment (HUmutOzdemir) [inline Platon/backend/app/search_engine/views.py:510]: There are some N/A's in the date parts of the Upcoming Events. So if a N/A comes to this part, the program will give an error. Please add an error handling here.

File: Platon/backend/app/search_engine/views.py [hunk 11]
```diff
@@ -433,7 +452,124 @@ def get(self):
...
+                        deadline_filter_end_list.append(result)
+                    result_list = deadline_filter_end_list
+            # Search tokens for date match
+            date_filter_start_list = []
+            if form.date_filter_start.data is not None:
+                for result in result_list:
+                    event_deadline = result.get("date")
+                    filter_data = form.date_filter_start.data
+                    if datetime.strptime(event_deadline[0:12].strip(), '%b %d, %Y') >= filter_data:
+                        date_filter_start_list.append(result)
```
Comment (HUmutOzdemir) [inline Platon/backend/app/search_engine/views.py:527]: Same as above

File: Platon/backend/app/search_engine/views.py [hunk 12]
```diff
@@ -433,7 +452,124 @@ def get(self):
...
+                        deadline_filter_start_list.append(result)
+                    result_list = deadline_filter_start_list
+            deadline_filter_end_list = []
+            if form.deadline_filter_end.data is not None:
+                for result in result_list:
+                    event_deadline = result.get("deadline")
+                    filter_data = form.deadline_filter_end.data
+                    if datetime.strptime(event_deadline, '%b %d, %Y') < filter_data:
+                        deadline_filter_end_list.append(result)
+                    result_list = deadline_filter_end_list
```
Comment (HUmutOzdemir) [inline Platon/backend/app/search_engine/views.py:519]: Same as above

File: Platon/backend/app/search_engine/views.py [hunk 13]
```diff
@@ -433,7 +452,124 @@ def get(self):
...
+                    if datetime.strptime(event_deadline[0:12].strip(), '%b %d, %Y') >= filter_data:
+                        date_filter_start_list.append(result)
+                    result_list = date_filter_start_list
+            date_filter_end_list = []
+            if form.date_filter_end.data is not None:
+                for result in result_list:
+                    event_deadline = result.get("date")
+                    filter_data = form.date_filter_end.data
+                    if datetime.strptime(event_deadline[0:12].strip(), '%b %d, %Y') < filter_data:
+                        date_filter_end_list.append(result)
```
Comment (HUmutOzdemir) [inline Platon/backend/app/search_engine/views.py:535]: Same as above
